drivers/i2s/i2s_litex.c

```diff
@@ -0,0 +1,617 @@
+ *
+ * @return i2s sample width in bits
+ */
+static uint32_t i2s_get_sample_width(int reg)
+{
+	uint32_t reg_data = litex_read32(reg + I2S_CONFIG_REG_OFFSET);
+	reg_data &= I2S_CONF_SAMPLE_WIDTH_MASK;
+	return reg_data >> I2S_CONF_SAMPLE_WIDTH_OFFSET;
+}
+
+/**
+ * @brief Get i2s audio sampling rate handled by device
+ *
+ * @param reg base register of device 
+ *
+ * @return audio sampling rate in Hz
+ */
+static uint32_t i2s_get_audio_freq(int reg)
+{
+	uint32_t reg_data = litex_read32(reg + I2S_CONFIG_REG_OFFSET);
+	reg_data &= I2S_CONF_LRCK_MASK;
+	return reg_data >> I2S_CONF_LRCK_FREQ_OFFSET;
+}
+
+/**
+ * @brief Enable i2s interrupt in event register
+ *
+ * @param reg base register of device
+ * @param irq_type irq type to be enabled one of I2S_EV_READY or I2S_EV_ERROR
+ *
+ * @return N/A
+ */
+static void i2s_irq_enable(int reg, int irq_type)
+{
+	assert(irq_type == I2S_EV_READY || irq_type == I2S_EV_ERROR);
+
+	uint8_t reg_data = litex_read8(reg + I2S_EV_ENABLE_REG_OFFSET);
+
+	litex_write8(reg_data | irq_type, reg + I2S_EV_ENABLE_REG_OFFSET);
+}
+
+/**
+ * @brief Disable i2s interrupt in event register
+ *
+ * @param reg base register of device
+ * @param irq_type irq type to be disabled one of I2S_EV_READY or I2S_EV_ERROR
+ *
+ * @return N/A
+ */
+static void i2s_irq_disable(int reg, int irq_type)
+{
+	assert(irq_type == I2S_EV_READY || irq_type == I2S_EV_ERROR);
+
+	uint8_t reg_data = litex_read8(reg + I2S_EV_ENABLE_REG_OFFSET);
+
+	litex_write8(reg_data & ~(irq_type), reg + I2S_EV_ENABLE_REG_OFFSET);
+}
+
+/**
+ * @brief Clear all pending irqs 
+ *
+ * @param reg base register of device 
+ *
+ * @return N/A
+ */
+static void i2s_clear_pending_irq(int reg)
+{
+	uint8_t reg_data = litex_read8(reg + I2S_EV_PENDING_REG_OFFSET);
+
+	sys_write8(reg_data, reg + I2S_EV_PENDING_REG_OFFSET);
+}
+
+/**
+ * @brief fast data copy function, 
+ * each operation copies 32 bit data chunks
+ * This function copies data from fifo into user buffer
+ *
+ * @param dst memory destination where fifo data will be copied to
+ * @param size amount of data to be copied
+ * @param sample_width width of signle sample in bits
+ * @param channels number of received channels
+ *
+ * @return N/A
+ */
+static void i2s_copy_from_fifo(uint8_t *dst, size_t size, int sample_width,
+			       int channels)
+{
+	uint32_t data;
+	int chan_size = sample_width / 8;
+#if CONFIG_I2S_LITEX_CHANNELS_CONCATENATED
+	if (channels == 2) {
+		for (size_t i = 0; i < size / chan_size; i += 4) {
+			//using sys_read function, as fifo is not a csr,
+			//but a contignous memory space
+			*(dst + i) = sys_read32(I2S_RX_FIFO_ADDR);
+		}
+	} else {
+		for (size_t i = 0; i < size / chan_size; i += 2) {
+			data = sys_read32(I2S_RX_FIFO_ADDR);
+			*((uint16_t*)(dst + i)) = data & 0xffff;
+		}
+	}
+#else
+	int max_off = chan_size - 1;
+	for (size_t i = 0; i < size / chan_size; ++i) {
+		data = sys_read32(I2S_RX_FIFO_ADDR);
+		for (int off = max_off; off >= 0; off--) {
```

Comment:
I'm not certain that Zephyr style allows defining variables in the for loop initialiser.  Lets see what others say.